api/grpcserver/v2alpha1/smeshing_identities.go

```diff
@@ -72,42 +73,97 @@ func (s *SmeshingIdentitiesService) States(
 
 		for i := len(history) - 1; i >= 0; i-- {
 			info := history[i]
-			ts := timestamppb.New(info.Time)
 			identityStateInfo := &pb.IdentityStateInfo{
-				State:   statusMap[info.State],
-				Time:    ts,
-				Message: info.Message,
+				State: statusMap[info.State],
+				Time:  timestamppb.New(info.Time),
 			}
 			if info.PublishEpoch != nil {
 				epoch := info.PublishEpoch.Uint32()
 				identityStateInfo.PublishEpoch = &epoch
 			}
+
+			if info.State == identity.StateRetrying && info.RetryingState != nil {
+				identityStateInfo.Metadata = &pb.IdentityStateInfo_Retrying{
+					Retrying: &pb.RetryingState{
+						Message: info.RetryingState.Error.Error(),
+					},
+				}
+			}
+
+			if info.State == identity.StatePoetRegistered && info.PoetRegisteredState != nil {
+				identityStateInfo.Metadata = &pb.IdentityStateInfo_PoetRegistered{
+					PoetRegistered: &pb.PoetRegisteredState{
+						Registrations: castRegistrations(info.PoetRegisteredState.Registrations),
+					},
+				}
+			}
+
+			if info.State == identity.StateWaitForPoetRoundEnd && info.WaitForPoetRoundEndState != nil {
+				identityStateInfo.Metadata = &pb.IdentityStateInfo_WaitForPoetRoundEnd{
+					WaitForPoetRoundEnd: &pb.WaitForPoetRoundEndState{
+						RoundEnd:        timestamppb.New(info.WaitForPoetRoundEndState.RoundEnd),
+						PublishEpochEnd: timestamppb.New(info.WaitForPoetRoundEndState.PublishEpochEnd),
+					},
+				}
+			}
+
+			if info.State == identity.StatePoetProofReceived && info.PoetProofReceivedState != nil {
+				identityStateInfo.Metadata = &pb.IdentityStateInfo_PoetProofReceived{
+					PoetProofReceived: &pb.PoetProofReceivedState{
+						PoetUrl: info.PoetProofReceivedState.PoetUrl,
+					},
+				}
+			}
+
+			if info.State == identity.StateATXBroadcasted && info.AtxBroadcastedState != nil {
+				identityStateInfo.Metadata = &pb.IdentityStateInfo_AtxBroadcasted{
+					AtxBroadcasted: &pb.AtxBroadcastedState{
+						AtxId: info.AtxBroadcastedState.AtxId.Bytes(),
+					},
+				}
+			}
+
+			if info.State == identity.StateProposalPublished && info.ProposalPublishedState != nil {
+				identityStateInfo.Metadata = &pb.IdentityStateInfo_ProposalPublished{
+					ProposalPublished: &pb.ProposalPublishedState{
+						Proposal: info.ProposalPublishedState.Proposal.Bytes(),
+						Layer:    info.ProposalPublishedState.Layer.Uint32(),
+					},
+				}
+			}
```

Comment:
Since `info.State` can only be one of these values, how about a `switch`/`case` here? 🙂